### DISCUSSION TITLE
REST Table Export

### DISCUSSION THREAD

**Discussion by aloreaux** [Budibase/budibase#7089] [Help] [answered]
Is there a way to export the table you get with REST API? I can convert the data to CSV format but unsure where to go to create the download button / link. Any help would be great.

**Comment by melohagan** [comment]
Hey @aloreaux 

There is an **Export Data** action that allows you to export rows from a table, but maybe that doesn't satisfy your use case?
![Screenshot 2022-08-08 at 12 43 29](https://user-images.githubusercontent.com/101575380/183410508-822f469e-297d-4e71-b959-bc21b9202dab.png)

**Reply by aloreaux** [reply]
So unfortunately it is looking to select rows to export but I have not been able to figure out how to do that. I tried but always got an error.

**Reply by melohagan** [reply]
Ah sorry I forgot that you cannot select rows for a REST query: https://github.com/Budibase/budibase/issues/4895#issuecomment-1157525518

There's no way I can think of to export the converted CSV API response. Depending on your dev experience, you could spin up a microservice or lambda function that supports the GET method and does the API call and returns the CSV file. 
Then it would just be a matter in Budibase of having a link with a URL that points to that service.

**Reply by aloreaux** [reply]
Hmm that is an interesting solution. I don't mind the coding aspect but I am unfamiliar with the microservice side of things. Is that just another docker container with some quick webserver to make all the calls?

**Reply by melohagan** [reply]
Yeah pretty much! Some common frameworks that make this easier include:
- Spring Boot (Java,Kotlin)
- Flask (Python)
- Express (NodeJS)

**Reply by aloreaux** [reply]
Thank you!

**Comment by aloreaux** [comment, marked as answer]
If anyone else is wondering about this. I used json2csv and express. Pointed my button to the link and it works great!

**Reply by Hedronmx** [reply]
Hey mind sharing how you did it exactly? Do you return the url or you return the csv data?

**Reply by aloreaux** [reply]
No problem. I used ExpressJS.

```
app.get('/getcsv', async (req, res) => {
  var token = await getToken();
  var opps = await getOpps(token);
  var csv = json2csvParser.parse(opps);
  res.setHeader('Content-Type', 'text/csv');
  res.setHeader('Content-Disposition', 'attachment; filename=output.csv');
});
```

So when I call https://someurl.com/getcsv it will prompt with a save dialog of where you would like to put it on your computer.

I then just add that url to a button in my budibase app and it allows the users to export to csv. To my knowledge it cannot be done directly in budibase, you need to setup an express environment (or alternative) to make the calls to get the data.

**Reply by tokkisva** [reply]
@aloreaux this is awesome, does your download button run the same GET as whatever bindings you have setup in BB, or is it a static GET?

I'm trying to do something similar where:
I have a form field in BB where a user inputs an id #, that field is attached to a binding, like {{formId}}.
When that ID# is filled out, BB runs GET {url}/someApi/{{Binding.formId}}

When the user clicks your download button, does your microservice run a GET with the binding value from a field like the example above?

**Reply by aloreaux** [reply]
@tokkisva There might be a way to do what you are talking about. You will probably have to pass it through the url. Currently mine is static and calls a GET from Express of the entire parsed data. I guess now that I think about it I am doing two GET's (BB & Express). I don't see why that wouldn't work I just didn't implement for my use case. Best of luck to you!

**Reply by aaakashjain** [reply]
Hi @aloreaux,

```
So when I call https://someurl.com/getcsv it will prompt with a save dialog of where you would like to put it on your computer.

I then just add that url to a button in my budibase app and it allows the users to export to csv.
```

I'm facing a file download issue when trying to download a file from a button on BB. 
I've created a GET URL (similar to /getcsv) that successfully downloads a CSV file when used in a browser. **However, when this URL is added to the button, the file doesn't download.** 

_Question: Is there any additional configuration required in the "Transformer" of the REST data source or any other action needed to enable file downloads from the button?_
